### PR TITLE
Flatening of the BVH

### PR DESCRIPTION
**Pull request by alexd2580**

### CODE REVIEW

File: src/flat_bvh.rs [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+/// let shapes = create_bounded_shapes();
+/// let bvh = BVH::build(&shapes);
+/// let flat_bvh = bvh.flatten();
+/// let hit_shape_indices = traverse_flat_bvh(&ray, &flat_bvh);
+/// ```
+pub fn traverse_flat_bvh(ray: &Ray, flat_nodes: &[FlatNode]) -> Vec<usize> {
+    let mut hit_shapes = Vec::new();
+    let mut index = 0;
+    let max_length = flat_nodes.len();
+    while index < max_length {
```
Comment (svenstaro) [inline src/flat_bvh.rs]: Please describe the logic and loop a little.

File: src/flat_bvh.rs [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+    /// Flattens the [`BVH`], so that it can be traversed in an iterative manner.
+    /// The iterative traverse procedure is implemented in [`traverse_flat_bvh()`].
+    ///
+    /// [`BVH`]: struct.BVH.html
+    /// [`traverse_flat_bvh()`]: method.traverse_flat_bvh.html
+    ///
+    pub fn flatten(&self, vec: &mut Vec<FlatNode>, next_free: usize) -> usize {
+        match *self {
+            BVHNode::Node { ref child_l_aabb, ref child_l, ref child_r_aabb, ref child_r } => {
+
```
Comment (svenstaro) [inline src/flat_bvh.rs]: Why this newline?

File: src/flat_bvh.rs [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+                    aabb: *child_l_aabb,
+                    entry_index: (next_free + 1) as u32,
+                    exit_index: 0,
+                    shape_index: u32::max_value(),
+                });
+
+                let index_after_child_l = child_l.flatten(vec, next_free + 1);
+                vec[next_free as usize].exit_index = index_after_child_l as u32;
+
+                vec.push(FlatNode {
```
Comment (svenstaro) [inline src/flat_bvh.rs]: Please add some comments around this function.

File: src/flat_bvh.rs [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+    /// [`traverse_flat_bvh`]: method.traverse_flat_bvh.html
+    ///
+    pub fn flatten_custom<F, FNodeType>(&self,
+                                        vec: &mut Vec<FNodeType>,
+                                        next_free: usize,
+                                        constructor: &F)
+                                        -> usize
+        where F: Fn(&AABB, u32, u32, u32) -> FNodeType
+    {
+
```
Comment (svenstaro) [inline src/flat_bvh.rs]: Why the spaces?

File: src/flat_bvh.rs [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+mod tests {
+    use aabb::{AABB, Bounded};
+    use bvh::BVH;
+    use bvh::tests::{XBox, build_some_bvh, create_n_cubes, create_ray};
+    use flat_bvh::{traverse_flat_bvh, FlatNode};
+    use nalgebra::{Point3, Vector3};
+    use ray::Ray;
+
+    #[test]
+    /// Tests whether the `flatten` procedure succeeds in not failing.
```
Comment (svenstaro) [inline src/flat_bvh.rs]: Describe what the test is trying to do here.

File: src/flat_bvh.rs [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+
+            // Traverse the resulting list of positive AABB tests
+            for index in &hits {
+                let triangle = &triangles[*index];
+                ray.intersects_triangle(&triangle.a, &triangle.b, &triangle.c);
+            }
+        });
+    }
+
+    #[test]
```
Comment (svenstaro) [inline src/flat_bvh.rs]: Put this up there with the other tests.

File: src/flat_bvh.rs [hunk 7]
```diff
@@ -0,0 +1,496 @@
...
+    /// Test whether the `flatten_custom` produces a flat bvh with the same relative structure
```
Comment (svenstaro) [inline src/flat_bvh.rs]: Capitalize BVH.

File: src/flat_bvh.rs (hunk 1 above)
Comment (alexd2580) [inline src/flat_bvh.rs]: Described in c636a98.

File: src/flat_bvh.rs [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+                }
+
+                next_shape
+            }
+        }
+    }
+}
+
+impl BVH {
+    /// Flattens the [`BVH`] so that it can be traversed iteratively.
```
Comment (alexd2580) [inline src/flat_bvh.rs]: Added in c636a98.

File: src/flat_bvh.rs [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+    ///
+    /// [`BVH`]: struct.BVH.html
+    ///
+    pub fn flatten(&self) -> Vec<FlatNode> {
+        let mut vec = Vec::new();
+        self.root.flatten(&mut vec, 0);
+        vec
+    }
+
+    /// Flattens the [`BVH`] so that it can be traversed iteratively.
```
Comment (alexd2580) [inline src/flat_bvh.rs]: Provided in c636a98.

File: src/flat_bvh.rs (hunk 3 above)
Comment (alexd2580) [inline src/flat_bvh.rs]: Added in 216fba8.

File: src/flat_bvh.rs (hunk 4 above)
Comment (alexd2580) [inline src/flat_bvh.rs]: No particular reason. Removed in 216fba8.

File: src/flat_bvh.rs (hunk 7 above)
Comment (alexd2580) [inline src/flat_bvh.rs]: Capitalized in 088a75d.

File: src/flat_bvh.rs (hunk 6 above)
Comment (alexd2580) [inline src/flat_bvh.rs]: Moved up in 088a75d.

File: src/flat_bvh.rs (hunk 5 above)
Comment (alexd2580) [inline src/flat_bvh.rs]: Described in 088a75d.

**Pull request merged**